### PR TITLE
one potential solution to envs/approaches __init__

### LINKED ISSUE #558: update envs/__init__.py and approaches/__init__.py to automatically discover modules
we'll need to make the names of envs/approaches, which are currently in the `__init__.py` files, instead be exposed via abstract methods in BaseEnv/BaseApproach

### PR DESCRIPTION
**Pull request by ronuchit**
This PR makes us not have to touch `__init__.py` in `approaches/` or `envs/` when adding new approaches or environments. One large change is that we can't import things directly from approaches or envs anymore, e.g.:
```
from predicators.src.envs import CoverEnv
from predicators.src.approaches import OracleApproach
```
now needs to be:
```
from predicators.src.envs.cover import CoverEnv
from predicators.src.approaches.oracle_approach import OracleApproach
```

closes #558

### CODE REVIEW

**Comment by ronuchit** [comment]
@tomsilver @williamshen-nz here's one potential solution, how do you feel about it

**Review by tomsilver — COMMENTED** [review]
I'm going to let @williamshen-nz take the lead on this one, but it looks reasonable to me at a high level

**Comment by ronuchit** [comment]
Verified that things work fine on supercloud with this change.

**Comment by williamshen-nz** [comment]
> This PR makes us not have to touch `__init__.py` in `approaches/` or `envs/` when adding new approaches or environments. One large change is that we can't import things directly from approaches or envs anymore, e.g.:
> 
> ```
> from predicators.src.envs import CoverEnv
> from predicators.src.approaches import OracleApproach
> ```
> 
> now needs to be:
> 
> ```
> from predicators.src.envs.cover import CoverEnv
> from predicators.src.approaches.oracle_approach import OracleApproach
> ```
> 
> closes #558

Hmm that's not ideal - doesn't it sort of defeat the purpose? I've worked on a similar feature in the past will have a closer look after lab meeting

**Comment by ronuchit** [comment]
Yeah, would be good to avoid that if we could... okay thanks!

File: src/approaches/__init__.py [hunk 1]
```diff
@@ -1,67 +1,33 @@
...
-    "BilevelPlanningApproach",
-    "NSRTLearningApproach",
-    "InteractiveLearningApproach",
-    "GrammarSearchInventionApproach",
-    "ApproachTimeout",
-    "ApproachFailure",
-]
+__all__ = ["BaseApproach", "ApproachTimeout", "ApproachFailure"]
+
+if not TYPE_CHECKING:
```
Comment (williamshen-nz) [inline src/approaches/__init__.py:14]: What happens if we do it if we're type checking? is it just slower?

**Review by williamshen-nz — APPROVED** [review]
🚢  Looks good to me! If we want to handle directly importing from `__init__` then I can look into it - let me know what you think. I'll see if there's any easy approach

File: src/approaches/__init__.py (hunk 1 above)
Comment (ronuchit) [reply src/approaches/__init__.py:14]: it crashes and burns horrendously LOL, mypy starts to search over all the modules that are imported in the whole codebase for some reason...

**Comment by ronuchit** [comment]
After discussing on slack, we agreed that we are okay with changing the importing mechanism, since that helps us avoid some annoying circular import issues.

**Pull request merged**